In [1]:
import pandas as pd

## Reading files

In [2]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [3]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [4]:
wh = pd.concat([wh2018,wh2019])

In [5]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [6]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

In [11]:
wh['year'].isna()

0      False
1      False
2      False
3      False
4      False
       ...  
151    False
152    False
153    False
154    False
155    False
Name: year, Length: 312, dtype: bool

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

In [17]:
tranform_to_string = ['year']
for c_ in tranform_to_string: wh[c_] = wh[c_].astype(str)
wh.dtypes

Overall rank                      int64
Country or region                object
Score                           float64
GDP per capita                  float64
Social support                  float64
Healthy life expectancy         float64
Freedom to make life choices    float64
Generosity                      float64
Perceptions of corruption       float64
year                             object
dtype: object

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

### resolucion ejercicio tres

In [64]:
wh['year'].isna()

0      False
1      False
2      False
3      False
4      False
       ...  
151    False
152    False
153    False
154    False
155    False
Name: year, Length: 312, dtype: bool

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [7]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

In [61]:
porcentaje = []

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro

In [62]:
for i in range(len(wh["Country or region"])):
    for j in range(len(wh["Country or region"])):
        if i != j:
              if wh["Country or region"][i] == wh["Country or region"][j] and wh["Country or region"][i] not in porcentaje:
                    ((porcentaje.append(wh["Country or region"][i]) == wh['Country or region'] - wh2019['Country or region']))

ValueError: Can only compare identically-labeled Series objects

In [63]:
print(porcentaje)

[]
